### PR TITLE
[Undertow] Support for enum values as query parameters

### PR DESCRIPTION
**Pull request by carterkozak**

### CODE REVIEW

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java
Comment (rubenfiszel) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java]: For once, I would argue that it might be more readable using an if-else statement:

`if (isPrimitive) ... else ...`

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java]: ha, updated

**Review by rubenfiszel — APPROVED** [review]

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java
Comment (iamdanfox) [inline conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java:571]: In conjure documentation, we've been using the term 'container' types to refer to list, set, map and optional.

https://palantir.github.io/conjure/#/docs/concepts?id=container-types

EDIT- actually where will 'map<K,V>' fit in to all this? Is this method just for generics with a single param?

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java:571]: Map is not allowed in query/path/header parameters

Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java:571]: This will never return a container type, just the contained type in the presence of a container, though an imported type is considered complex, and we use the valueOf function on it for parity with jersey.
I should clearly write some javadoc :-)

File: conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/internal/StringDeserializers.java
Comment (uschi2000) [inline conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/internal/StringDeserializers.java:465]: should add `@Nullable` (same above)

File: conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java
Comment (carterkozak) [reply conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java:571]: Updated

File: conjure-java-core/src/test/resources/ete-service.yml
Comment (uschi2000) [inline conjure-java-core/src/test/resources/ete-service.yml:152]: interestingly, the wire format spec doesn't allow this: `If the param-type is path, header or query then the de-aliased type MUST be an enum, a primitive or an optional of one of these. `

https://palantir.github.io/conjure/#/docs/spec/conjure_definitions

File: conjure-java-core/src/integrationInput/java/com/palantir/product/EteServiceRetrofit.java
Comment (iamdanfox) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/EteServiceRetrofit.java:127]: I don't think we were intending to support Lists or Sets in query / header paramters?

> If the param-type is path, header or query then the de-aliased type MUST be an enum, a primitive or an optional of one of these. Additionally, the type MUST NOT be bearertoken when the param-type is header or query

https://palantir.github.io/conjure/#/docs/spec/conjure_definitions

**Comment by uschi2000** [comment]
code looks fine, and yet: we need to either update the language in the spec (https://palantir.github.io/conjure/#/docs/spec/conjure_definitions) to allow list<enum> and add conjure-verification tests, or ban these constructs in the conjure yaml.

File: conjure-java-core/src/integrationInput/java/com/palantir/product/EteServiceRetrofit.java
Comment (carterkozak) [reply conjure-java-core/src/integrationInput/java/com/palantir/product/EteServiceRetrofit.java:127]: We already support both lists and sets in query parameters, we use them heavily across the board, this provides parity with the jersey implementation.

Happy to update the spec, but I'd prefer not to block this change on it.

File: conjure-java-core/src/integrationInput/java/com/palantir/product/EteService.java
Comment (iamdanfox) [inline conjure-java-core/src/integrationInput/java/com/palantir/product/EteService.java:168]: Rather than having to build up test cases in this EteService.java (which only benefits this one particular project) I'd prefer to improve the test cases in the conjure-verification project: https://github.com/palantir/conjure-verification/pull/109

This ensures that all language implementations will be forced to handle enums as path, query, header params!

File: conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/internal/StringDeserializers.java
Comment (carterkozak) [reply conjure-undertow-lib/src/main/java/com/palantir/conjure/java/undertow/lib/internal/StringDeserializers.java:465]: Good call, fixed!

**Comment by carterkozak** [comment]
PR: https://github.com/palantir/conjure/pull/191

File: conjure-java-core/src/test/resources/ete-service.yml
Comment (carterkozak) [reply conjure-java-core/src/test/resources/ete-service.yml:152]: Opened palantir/conjure#191

**Review by iamdanfox — COMMENTED** [review]
I'd like to get the new conjure-verification 0.16.5 tests passing here.  Don't think we should block this PR if it turns out to be a much larger effort to make conjure-verification support lists of enums as query parameters.

**Comment by carterkozak** [comment]
+1 for merging this to unblock verifier fixes -- I follow up once the verifier has been upgraded to deduplicate cases from the undertow ete test service.

**Comment by iamdanfox** [comment]
@cakofony can you cherry-pick 14a477f? I can't push to your branch.

Separately, I think it would be great to the conjure-verification tests actually running against some of the generated undertow handlers!!

**Comment by carterkozak** [comment]
> can you cherry-pick 14a477f? I can't push to your branch.

Currently all java implementations will fail the negative test on "!!!" because it's not rejected, but used as a valid UNKNOWN string value

**Comment by carterkozak** [comment]
I was planning to fix enum validation on a separate branch because it's not related to undertow

**Comment by iamdanfox** [comment]
OK fair enough - I tried to upgrade to 0.16.5 independently from this PR but the undertow generator threw an error as part of the pre-test codegen!

**Review by iamdanfox — APPROVED** [review]

**Pull request merged**
